crates/p2p/src/client/peer_agnostic.rs

```diff
@@ -102,6 +102,87 @@ impl Client {
         peers
     }
 
+    pub async fn header_stream(
+        self,
+        start: BlockNumber,
+        stop: BlockNumber,
+        reverse: bool,
+    ) -> impl futures::Stream<Item = pathfinder_common::SignedBlockHeader> {
+        let (mut start, stop, direction) = match reverse {
+            true => (stop, start, Direction::Backward),
+            false => (start, stop, Direction::Forward),
+        };
+
+        async_stream::stream! {
+            // Loop which refreshes peer set once we exhaust it.
+            loop {
+                let peers = self
+                    .get_update_peers_with_sync_capability(protocol::Headers::NAME)
+                    .await;
+
+                // Attempt each peer.
+                'next_peer: for peer in peers {
```

Comment:
block numbers are already verified, would be nice to check for block hash duplication too :pray:

Reply:
Block hash duplication? The caller will have to verify / check block hashes anyway, so storing all block hashes just to check for duplicates doesn't seem useful?